File: No_1411/Number of Ways to Paint N × 3 Grid.cpp
class Solution {
public:
    int numOfWays(int n) {
        long long pattern1 = 6, pattern2 = 6;
        const int MOD = 1e9 + 7;
        n--;
        while (n--) {
            long long t = pattern1;
            pattern1 = (pattern1 * 2 + pattern2 * 2) % MOD;
            pattern2 = (pattern2 * 3 + t * 2) % MOD;
        }
        
        return (pattern1 + pattern2) % MOD;
    }     
};

File: No_1411/Number of Ways to Paint N × 3 Grid.py
class Solution:
    def numOfWays(self, n: int) -> int:
        pattern1 = 6  # c1 c2 c3
        pattern2 = 6  # c1 c2 c1
        MOD = 10**9 + 7
        for _ in range(n - 1):
            new_pattern1 = (2 * pattern1 + 2 * pattern2) % MOD
            new_pattern2 = (2 * pattern1 + 3 * pattern2) % MOD
            pattern1, pattern2 = new_pattern1, new_pattern2
        
        return (pattern1 + pattern2) % MOD
            